lib/ClangImporter/ClangImporter.cpp

```diff
@@ -4747,6 +4736,23 @@ Decl *ClangImporter::importDeclDirectly(const clang::NamedDecl *decl) {
   return Impl.importDecl(decl, Impl.CurrentVersion);
 }
 
-void ClangImporter::diagnoseDeclDirectly(const clang::NamedDecl *decl) {
-  Impl.diagnoseTargetDirectly(decl);
+void ClangImporter::diagnoseTopLevelValue(const DeclName &name) {
+  Impl.diagnoseTopLevelValue(name);
+}
+
+void ClangImporter::diagnoseMemberValue(const DeclName &name,
+                                        const Type &baseType) {
+  if (!baseType->getAnyNominal())
```

Comment:
Couple changes here. Since this call is made from Sema, the baseType could be anything (not just an imported ClangType). I discovered a few more checks are needed here. 

First, `extractDirectlyReferencedNominalTypes` only knows how to handle small set of types of types. If you pass something it isn't expecting, we hit an `llvm_unreachable`. The `getAnyNominal` type takes care of this (I figure all imported types referencable from swift are nominal). 

Lastly, I added checks that `getClangDecl()` returns a non-null value, and that this value is a `clang::DeclContext`. If either of these checks fail, but the value is passed to `cast<clang::DeclContext>()`, it'll throw.